Core/src/IfcGeometryConverter/FaceConverter.h

```diff
@@ -1282,92 +1282,113 @@ namespace OpenInfraPlatform {
+								carve::math::Matrix planeMatrix = GeomUtils::convertPlane2Matrix(normalOfPlane, outerLoop[0],
+									outerLoop[1] - outerLoop[0]); 
 
 								carve::math::Matrix inversePlaneMatrix = GeomUtils::computeInverse(planeMatrix);
 
-
-								for (int i = 0; i < loops.size(); i++) 
+								for (const auto& loop : loops) 
 								{
-									std::vector<carve::geom::vector<3>> loop = loops[i];
 									std::vector<carve::geom::vector<2>> loop2D;
-									for (int j = 0; j < loop.size(); j++)
+									std::vector<carve::geom::vector<3>> loop3Dto2D;
+									for (const auto& point : loop)
 									{
-										carve::geom::vector<3> point3Dto2D = inversePlaneMatrix * loop[j];
-										
-										carve::geom::vector<2> point2D = carve::geom::VECTOR(point3Dto2D.x, point3Dto2D.y);
-										loop2D.push_back(point2D);
+										carve::geom::vector<3> point3Dto2D = inversePlaneMatrix * point;
+										loop3Dto2D.push_back(point3Dto2D);
 									}
+									loop2D = removeEmptyCoordinate(loop3Dto2D);
 									loops2D.push_back(loop2D);
 								}
 								
 								// 2. Call geomUtils
-								/*carve::geom::vector<3> normal_first_loop;
-								std::vector<std::vector<carve::geom2d::P2>>	face_loops = GeomUtils::correctWinding(loops, normal_first_loop);
+								carve::geom::vector<3> normalFirstLoop;
+								std::vector<std::vector<carve::geom2d::P2>>	faceLoops = GeomUtils::correctWinding(loops2D, normalFirstLoop);
+
+								std::vector<carve::geom2d::P2> mergedPath;
+								std::vector<carve::triangulate::tri_idx> triangulatedList;
+								std::vector<std::pair<size_t, size_t>> pathAllLoops;
 
-								std::vector<carve::geom2d::P2> merged_path;
-								std::vector<carve::triangulate::tri_idx> triangulated;
-								std::vector<std::pair<size_t, size_t>> path_all_loops;
+								GeomUtils::incorporateVoids(faceLoops, mergedPath, triangulatedList, pathAllLoops);
 
-								GeomUtils::incorporateVoids(face_loops, merged_path, triangulated, path_all_loops);*/
 								// 3. Given the results, find corresponding points in the original array. 
-								
+								std::vector<size_t> indexedPath;
+								for (const auto& path : mergedPath)
+								{
+									int i = 0;
+									for (const auto& loop2D : loops2D)
+									{
+										std::vector<size_t> indexLoop = indexLoops[i];
+										int j = 0;
+										for (const auto& point2D : loop2D)
+										{
+											size_t index = indexLoop[j];
+											carve::geom2d::P2 pointInPath(point2D);
+											if (path == pointInPath) 
+											{
+												indexedPath.push_back(index);
+											}
+											j++;
+										}
+										i++;
+									}
+								}
+
+								for (const auto& triangulated : triangulatedList)
+								{
+									int a = triangulated.a;
+									int b = triangulated.b;
+									int c = triangulated.c;
+
+									polygon->addFace(indexedPath[a], indexedPath[b], indexedPath[c]);
+								}

```

Comment:
This seems OK, but I'm unsure if it really works - we need to discuss this.

Reply:
Resolved:
9cc48d8